### DISCUSSION TITLE
VS 2022 Local Debug Durable Functions Issues

### DISCUSSION THREAD

**Discussion by ahawes-clarity** [Azure/azure-functions-durable-extension#2025] [Q&A] [answered]
I have an Azure Functions v3 application that I have been developing and running for several years. In VS 2019 I could debug locally without issue. Today when I opened the application in VS 2022 and tried to debug I ran into several issues.

There was one issue related to AzureWebJobsSecretStorageType that I solved by adding "AzureWebJobsSecretStorageType": "files" to my local.settings.json based on this post.
https://stackoverflow.com/questions/67728499/azure-functions-there-was-an-error-performing-a-read-operation-on-the-blob-stor

But after that I am still seeing errors that seem related to Durable Functions and I am not finding any answers online so far.

One issue is a timer function listener that triggers a Durable Functions clean up is unable to start due to Azure.Storage.Blobs: Service request failed. But then right after it seems to be able to start just fine. Why would it not start and then start?

Second is an "Error in LeaseManagerStarter task. Exception: Microsoft.WindowsAzure.Storage.StorageException: The response ended prematurely, with at least 128 additional bytes expected." which then leads to a "Error in LeaseManagerStarter task. Exception: System.InvalidOperationException: AppLeaseManager has already started"

I suspect maybe an issue with Azure Storage Emulator because previously in VS2019 when I would debug the Functions app it would automatically start Azure Storage Emulator but I don't see that happening, but maybe it happens a different way in VS2022. I do have "AzureWebJobsStorage": "UseDevelopmentStorage=true" in settings.

Here is what I see in the console:
![image](https://user-images.githubusercontent.com/6013584/144080126-c374b952-5723-4cb4-ac1a-5d0a4cd5fafd.png)

**Comment by cgillum** [comment, marked as answer]
It looks like you’re running Azurite and not the older Azure Storage Emulator. It also looks like you’re running a much newer version of the Functions runtime. I can tell because I see the newer Track 2 Azure Storage SDKs in these error messages. All of these changes to your environment are likely related to the problems you’re seeing.

Can you confirm the version of the Core Tools that gets run when you start debugging?

@fabiocav  are there any known compatibility issues with Functions V4 core tools and Azurite that might explain this?

**Reply by ahawes-clarity** [reply]
Azure Functions Core Tools
Core Tools Version:       3.0.3904 Commit hash: c345f7140a8f968c5dbc621f8a8374d8e3234206  (64-bit)
Function Runtime Version: 3.3.1.0

**Reply by ahawes-clarity** [reply]
Is there any update on this?

Was there something I should have done differently? Is there a migration guide I should know about? I didn't choose Azurite so it must just be something that came as part VS2022. My guess the expectation is that it should just work.

**Reply by cgillum** [reply]
Do you have the Azure Storage emulator also installed on your machine (usually at `%PROGRAMFILES(x86)%\Microsoft SDKs\Azure\Storage Emulator\AzureStorageEmulator.exe`)? If so, could you try running the Azure Storage Emulator before you start debugging to see if that resolves the issue? If the storage emulator is already running then Azurite won't be started up. I'm wondering if this could be a compatibility issue with Azurite.

**Reply by ahawes-clarity** [reply]
Yes, Azure Storage emulator is installed on my development machine.

If I startup Azure Storage emulator manually before debugging in VS2022 then it does seem to run as expected, probably because, as you noted, "Azurite won't be started up."

So, it would seem, this may indeed be a compatibility issue with Azurite as you suggest.

**Reply by ahawes-clarity** [reply]
I have been starting up Azure Storage Emulator before debugging and it was working fine until a recent VS update I think. I don't have to do this work all the time so I am not exactly sure what VS update may have been the cause.

This all just seems like a sub-optimal user experience.

Any idea when this might be addressed?

**Reply by ahawes-clarity** [reply]
> I have been starting up Azure Storage Emulator before debugging and it was working fine until a recent VS update I think. I don't have to do this work all the time so I am not exactly sure what VS update may have been the cause.
> 
> This all just seems like a sub-optimal user experience.
> 
> Any idea when this might be addressed?

It seems the issue was with Azure Storage Emulator not starting up. Not sure what broke it but re-installing (repairing) my Azure Storage Explorer install got things back working again. Still, I wish there was a way to let VS just start up Azure Storage Explorer on its own instead of using Azurite or making Azurite just work.

**Reply by davidmrdavid** [reply]
@ahawes-clarity, in your last response, when you said "Azure Storage Explorer", did you mean to say "Azure Storage Emulator"? Just asking since those two are different components and I would be surprised to hear the issue was in the explorer.

As for when this will be addressed, unfortunately we (the Durable Functions team) do not own this component directly so we can't say with certainty. After getting more clarity about whether this issue was in the explorer or not, I can try to find the owners and ask them to chime in.

**Reply by ahawes-clarity** [reply]
@davidmrdavid, Yes, sorry. I will try to update it. Thanks

**Reply by davidmrdavid** [reply]
Thanks. I'm reaching out internally to find the right product owners. Seems like a VisualStudio integration feature

**Reply by davidmrdavid** [reply]
@ahawes-clarity, I just spoke with an owner of this feature. They tell me that, in VS2022, they dropped support for the Azure Storage Emulator **in favor of** Azurite.

So I need a few things clarified.
1. Is this issue referring to VS2022? The title says VS2020 but in the thread I see references to VS2022 so I'm not certain.
2. Do you have Azurite installed?
3. If you do not have Azurite installed, can you try installing it and then seeing if this error continues?

Thanks! Please tag me in your answer so I can follow-up with feature owner.

**Comment by ahawes-clarity** [comment]
@davidmrdavid 
Thanks for looking into this.
1. Yes VS 2022
2. I was told Azurite installs automatically with VS 2022 and is used by default unless Azure Storage Emulator is loaded first.
3. Can you confirm I need to install it separately?

Heading out for the weekend now.

**Reply by davidmrdavid** [reply]
Thanks @ahawes-clarity.

You were right, it seems VS2022 installs Azurite by default, and should also start it up by default. 
With that in mind, I've been asked to request that you provide us with the contents of the "Service dependencies" pane in the output window. That should have the information we need to debug this. At your earliest convenience, please provide us with that data. Thanks!

**Reply by ahawes-clarity** [reply]
Thanks for helping.

Just after loading the solution the Service Dependencies output says this:
![image](https://user-images.githubusercontent.com/6013584/165094067-b7d96dae-dafc-411f-9d5c-de68c8ecf316.png)

Here is a screenshot of the errors as I see them now which seems to be the same as when I first posted. I don't get these errors if I launch Azure Storage Emulator before running the solution.
![image](https://user-images.githubusercontent.com/6013584/165095010-c302a69a-519b-4d0c-903b-5632ae6273bf.png)

**Reply by davidmrdavid** [reply]
Will reply back once I hear from feature owners, thanks for sharing this info :-)

**Reply by vijayrkn** [reply]
@davidmrdavid - This means the Azurite is starting and running as expected. What ever error we are getting is from the Functions not being able to use the Azurite.

**Reply by davidmrdavid** [reply]
Hmm, I use Azurite on a daily basis so I suspect something else is amiss, but if the logs indicate Azurite is working fine, then it makes sense to start looking elsewhere.

@ahawes-clarity: do you get those errors if you start *Azurite* (not the storage emulator, whose usage is deprecated) before VS2022? Also, do you have a small repro project you could zip into this thread for me to reproduce this error?

**Comment by ahawes-clarity** [comment]
@davidmrdavid  The problem started when I moved the solution to VS 2022. The project works fine in VS 2019. I understand that is because VS 2019 uses the Azure Storage Emulator and VS 2022 uses Azurite. So, either VS 2022 should just work or there should be some migration steps documented someplace on how to convert a project from Azure Storage Emulator to Azurite.

This is what I see if I try to launch Azurite via the command line:
![image](https://user-images.githubusercontent.com/6013584/165173983-6ddd1c4d-98ef-47a7-8316-a6b65000b89e.png)

I do not have a small repro project that I can provide. All my time is billable to my client so I can't justify spending extra time to create a small repro project. As it is, I am stretching it just spending the time to find a solution. And the only reason I am doing this is because launching Azure Storage Emulator before running the solution is a pain and I really would like not to be using something Microsoft has deprecated.

I can provide the whole solution privately but it is relatively complex.

**Reply by davidmrdavid** [reply]
Thanks @ahawes-clarity, I understand what you're saying and I agree that there ought to be good instructions about this, and there might be. From the Durable Functions perspective, we're a bit limited by how much we can help as this is not a component we own, so I'm trying to determine who is best to help you. Looking at the traces you've shared, and our discussion so far, this does not appear to be either at the Durable Functions level or the Azure Functions level - I think this is most likely on the Azurite front. 

@vijayrkn: do you know who would be a good point of contact to help debug on the Azurite front? This is not a problem on the Functions side

**Reply by davidmrdavid** [reply]
@ahawes-clarity: do you mind if we transfer this ticket to https://github.com/Azure/Azurite? I think they'd be better equipped to help

**Reply by vijayrkn** [reply]
ya, Moving to that repo would help. Thanks!

**Reply by ahawes-clarity** [reply]
Fine with me.

**Reply by davidmrdavid** [reply]
Ugh, I just remembered this is a "Discussions" thread, not an Issue thread, which prevents me from transferring.
@ahawes-clarity: To get support from the Azurite folks, we'll need you to create a new issue thread here: https://github.com/Azure/Azurite/issues/new

I could create an issue there myself, but I wouldn't be able to fill their issue-creation form with your data.
Please do feel free to tag me in there, and definitely tag this issue. Thanks!

**Reply by ahawes-clarity** [reply]
@davidmrdavid: Posted the issue here:
https://github.com/Azure/Azurite/issues/1447

**Reply by davidmrdavid** [reply]
Thanks @ahawes-clarity . I'll let the Azurite folks take over this investigation as they seem like the appropriate feature owners here. Do reach out if you need involvement from the Durable Functions team again, thanks and apologies for the inconveniences you're experiencing.

**Comment by ahawes-clarity** [comment]
I will try to do this tomorrow. Thanks
